File: plugin-comms-twosix-java/plugin/src/main/java/race/TwosixWhiteboardLinkProfileParser.java

// 
// Copyright 2023 Two Six Technologies
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// 

package race;

import ShimsJava.JLinkConfig;
import ShimsJava.RaceLog;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

public class TwosixWhiteboardLinkProfileParser extends LinkProfileParser {

    static final String logLabel = "race.TwosixWhiteboardLinkProfileParser";
    private String hostname = "";
    private int port = 0;
    private String hashtag = "";
    private int checkFrequency = 1000;
    private double timestamp = -1.0;

    /**
     * @param original
     * @return String
     */
    String fixHastag(String original) {
        String fixed = original;
        fixed.replaceAll("([^a-z0-9])", "");
        if (fixed.compareTo(original) != 0) {
            RaceLog.logWarning(
                    logLabel,
                    "Warning: the orignal hashtag contained invalid characters, fixed to " + fixed,
                    "");
        }
        return fixed;
    }

    TwosixWhiteboardLinkProfileParser(String linkProfile) {
        JSONParser parser = new JSONParser();
        JSONObject linkProfileJson = null;

        try {
            linkProfileJson = (JSONObject) parser.parse(linkProfile);
        } catch (Exception e) {
            RaceLog.logError(logLabel, "Invalid link profile (malformed JSON): " + linkProfile, "");
        }

        // required
        if (linkProfileJson.get("hostname") != null) {
            hostname = (String) linkProfileJson.get("hostname");
        } else {
            RaceLog.logError(
                    logLabel, "Invalid link profile (missing hostname): " + linkProfile, "");
        }
        // required
        if (linkProfileJson.get("port") != null) {
            try {
                port = Math.toIntExact((Long) linkProfileJson.get("port"));
            } catch (Exception e) {
                RaceLog.logError(logLabel, e.getMessage(), "");
            }

        } else {
            RaceLog.logError(logLabel, "Invalid link profile (missing port): " + linkProfile, "");
        }

        // required
        if (linkProfileJson.get("hashtag") != null) {
            hashtag = (String) linkProfileJson.get("hashtag");
            hashtag = fixHastag(hashtag);
        } else {
            RaceLog.logError(
                    logLabel, "Invalid link profile (missing hashtag): " + linkProfile, "");
        }

        // optional
        if (linkProfileJson.get("checkFrequency") != null) {
            checkFrequency = Math.toIntExact((Long) linkProfileJson.get("checkFrequency"));
        }

        // optional
        if (linkProfileJson.get("timestamp") != null) {
            timestamp = ((Number) linkProfileJson.get("timestamp")).doubleValue();
        }
    }

    /**
     * @param plugin
     * @param linkConfig
     * @return Link
     */
    @Override
    Link createLink(PluginCommsTwoSixJava plugin, JLinkConfig linkConfig, String channelGid) {
        String linkId = plugin.getjSdk().generateLinkId(channelGid);
        if (linkId == null || linkId.trim().isEmpty()) {
            RaceLog.logError(logLabel, "received invalid link ID from the SDK.", "");
            return null;
        }

        RaceLog.logDebug(logLabel, "Creating Twosix Whiteboard Link: " + linkId, "");
        return new TwosixWhiteboardLink(
                plugin,
                linkId,
                linkConfig.linkProfile,
                linkConfig.personas,
                linkConfig.linkProps,
                hostname,
                port,
                hashtag,
                checkFrequency,
                checkFrequency,
                timestamp);
    }
}

File: plugin-comms-twosix-cpp/source/whiteboard/TwosixWhiteboardLinkProfileParser.cpp

//
// Copyright 2023 Two Six Technologies
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "TwosixWhiteboardLinkProfileParser.h"

#include <nlohmann/json.hpp>
#include <regex>

#include "../PluginCommsTwoSixCpp.h"
#include "../config/LinkConfig.h"
#include "../utils/log.h"
#include "TwosixWhiteboardLink.h"

using json = nlohmann::json;

std::string TwosixWhiteboardLinkProfileParser::fixHashtag(const std::string &orig) {
    // Allow alphanumeric upper and lower case, underscores, and dashes
    std::regex e("([^a-zA-Z0-9_\\-])");
    std::string fixed = std::regex_replace(orig, e, "");
    if (fixed.compare(orig) != 0) {
        logWarning("Warning: the original hashtag \"" + orig +
                   "\" contained invalid characters, fixed to " + fixed);
    }
    return fixed;
}

TwosixWhiteboardLinkProfileParser::TwosixWhiteboardLinkProfileParser(
    const std::string &linkProfile) :
    LinkProfileParser(linkProfile) {
    json linkProfileJson;
    try {
        linkProfileJson = json::parse(linkProfile);
        logDebug("TwosixWhiteboardLinkProfileParser: link profile: " + linkProfile);

        // required
        hostname = linkProfileJson.at("hostname").get<std::string>();
        port = linkProfileJson.at("port").get<int>();

        // fixHashtag helps prevent stupid mistakes like putting / in the tag, which would break
        // accessing the whiteboard. Carried over from pixelfed. Better safe than sorry.
        hashtag = fixHashtag(linkProfileJson.at("hashtag").get<std::string>());

        // optional, but must be of correct type if it exists
        checkFrequency = linkProfileJson.value("checkFrequency", 1000);

        maxTries = linkProfileJson.value("maxTries", 120);
        logDebug("TwosixWhiteboardLinkProfileParser: maxTries: " + std::to_string(maxTries));

        timestamp = linkProfileJson.value("timestamp", -1.0);
    } catch (std::exception &error) {
        logError("TwosixWhiteboardLinkProfileParser: failed to parse link profile: " +
                 std::string(error.what()));
        logError("TwosixWhiteboardLinkProfileParser: invalid link profile: " + linkProfile);
        throw std::invalid_argument("invalid link profile");
    }
}

std::shared_ptr<Link> TwosixWhiteboardLinkProfileParser::createLink(IRaceSdkComms *sdk,
                                                                    PluginCommsTwoSixCpp *plugin,
                                                                    Channel *channel,
                                                                    const LinkConfig &linkConfig,
                                                                    const std::string &channelGid) {
    LinkID linkId = sdk->generateLinkId(channelGid);

    logDebug("Creating Twosix Whiteboard Link: " + linkId);
    return std::make_shared<TwosixWhiteboardLink>(sdk, plugin, channel, linkId,
                                                  linkConfig.linkProps, *this);
}
